src/storage/hummock_trace/src/collector.rs

```diff
@@ -0,0 +1,408 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::env;
+use std::fs::{create_dir_all, OpenOptions};
+use std::io::BufWriter;
+use std::path::Path;
+use std::sync::atomic::AtomicU64;
+
+use bincode::{Decode, Encode};
+use either::Either;
+use flume::{unbounded, Receiver, Sender};
+use tokio::task_local;
+
+use crate::write::{TraceWriter, TraceWriterImpl};
+use crate::{
+    ConcurrentIdGenerator, Operation, OperationResult, Record, RecordId, RecordIdGenerator,
+    UniqueIdGenerator,
+};
+
+// create a global singleton of collector as well as record id generator
+lazy_static! {
+    static ref GLOBAL_COLLECTOR: GlobalCollector = GlobalCollector::new();
+    static ref GLOBAL_RECORD_ID: RecordIdGenerator = UniqueIdGenerator::new(AtomicU64::new(0));
+    static ref SHOULD_USE_TRACE: bool = set_use_trace();
+    pub static ref CONCURRENT_ID: ConcurrentIdGenerator = UniqueIdGenerator::new(AtomicU64::new(0));
+}
+
+const USE_TRACE: &str = "USE_HM_TRACE";
+const LOG_PATH: &str = "HM_TRACE_PATH";
+const DEFAULT_PATH: &str = ".trace/hummock.ht";
+const WRITER_BUFFER_SIZE: usize = 1024;
+
+pub fn should_use_trace() -> bool {
+    *SHOULD_USE_TRACE
+}
+
+fn set_use_trace() -> bool {
+    if let Ok(v) = std::env::var(USE_TRACE) {
+        v.parse().unwrap()
+    } else {
+        false
+    }
+}
+
+/// Initialize the `GLOBAL_COLLECTOR` with configured log file
+pub fn init_collector() {
+    let path = match env::var(LOG_PATH) {
+        Ok(p) => p,
+        Err(_) => DEFAULT_PATH.to_string(),
+    };
+    let path = Path::new(&path);
+
+    if let Some(parent) = path.parent() {
+        if !parent.exists() {
+            create_dir_all(parent).unwrap();
+        }
+    }
+
+    let f = OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .create(true)
+        .open(path)
+        .expect("failed to open log file");
+
+    let writer = BufWriter::with_capacity(WRITER_BUFFER_SIZE, f);
+    let writer = TraceWriterImpl::new_bincode(writer).unwrap();
+    tokio::spawn(GLOBAL_COLLECTOR.run(Box::new(writer)));
+}
+
+/// `GlobalCollector` collects traced hummock operations.
+/// It starts a collector thread and writer thread.
+struct GlobalCollector {
+    tx: Sender<RecordMsg>,
+    rx: Receiver<RecordMsg>,
+}
+
+impl GlobalCollector {
+    fn new() -> Self {
+        let (tx, rx) = unbounded();
+        Self { tx, rx }
+    }
+
+    async fn run(&self, writer: Box<dyn TraceWriter + Send>) {
```

Comment:
Since this is a future to be spawed, for the writer we can set its type as `writer: impl TraceWriter + Send`, and then the type of writer can be determined statically at compile time, and we can avoid looking up the vtable and do dynamic dispatch with Box<dyn ...>.